Voyager verification support

ametel01 opened the pull request:
reopened after force push
Nethermind's rpc url has been removed because discontinued.

ametel01 committed:
trigger reopen of voyager-verify PR

ametel01 committed:
Adds Voyager verifier

Enables contract verification using the Voyager block explorer.
Introduces a new `Voyager` struct implementing the `VerificationInterface`.
The implementation gathers project source files, constructs a payload, and submits it to the Voyager API for verification.

Also, some refactoring for the verification command has been done:
- Added `rpc` param to allow overriding `url` from `snfoundry.toml`
- Added `Voyager` option in `Verifier` enum
- Provider moved to `VerificationInterface`

ametel01 committed:
Simplifies RPC URL generation for networks

Removes the `FreeProvider` abstraction and directly constructs RPC URLs for mainnet and sepolia using Blast endpoints. This simplifies the URL generation process and removes the need for semi-random provider selection.

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Simplifies RPC provider selection

Removes the semi-random RPC provider selection logic, using Blast as the default provider.

This change simplifies the provider selection process by removing the reliance on installation-dependent seeds, ensuring a consistent provider across different environments.

ametel01 committed:
Updates Walnut verification response handling

Modifies the Walnut verification process to handle the verification response as a success, regardless of the HTTP status code. Updates tests to reflect the new response message from the verifier.

ametel01 committed:
Adds Voyager verifier tests

Adds end-to-end tests for the Voyager verifier to ensure it functions correctly in different scenarios, including successful and failed verifications, handling contract address and class hash inputs, and virtual workspaces.

ametel01 committed:
Updates contract verification tests

Updates contract verification tests to align with changes in verification process.

The verifier response is simplified to "Contract successfully verified".
The environment variable name is updated from `WALNUT_API_URL` to `VERIFIER_API_URL`.

ametel01 committed:
Updates test setup for RPC call parameters

Updates the test setup to align with the expected RPC call parameters format by using a structured object instead of a list for the `starknet_getClassHashAt` method.

Additionally, changes `assert_stderr_contains` to `assert_stdout_contains` and changes wording to match.

ametel01 committed:
Improves `sncast verify` functionality

Adds support for Voyager block explorer verification.

Removes the default usage of Walnut for `sncast verify`, requiring explicit specification.

Updates OpenZeppelin account to v1.0.0 preset and restores support for Braavos accounts with updated class hash.

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Filters out test files from source packaging

Ensures that test files are not included when packaging source files for verification.
This prevents issues that may arise from including test-related code or attributes.
It excludes files located in "tests" directories and files containing the "#[test]" attribute.

franciszekjob commented:
Works fine 👍 
https://sepolia.voyager.online/class/0x07cbbbb8b0ad2fe495432dd947c830d02b1c7431c9f80f45f02607b61d5cbd68
<img width="298" alt="image" src="https://github.com/user-attachments/assets/fa2277c0-4f0e-45cd-b520-cdac29ed48e2" />

ametel01 committed:
addressed review comments

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
fixed wrong cli arg on voyager e2e tests

franciszekjob commented:
@ametel01 some tests are still failing
https://github.com/foundry-rs/starknet-foundry/actions/runs/15569474990/job/43841653996?pr=3445

ametel01 commented:
> @ametel01 some tests are still failing
> 
> https://github.com/foundry-rs/starknet-foundry/actions/runs/15569474990/job/43841653996?pr=3445

Fixing, switching to foundry UI seems to have caused this

franciszekjob commented:
> > @ametel01 some tests are still failing
> > https://github.com/foundry-rs/starknet-foundry/actions/runs/15569474990/job/43841653996?pr=3445
> 
> Fixing, switching to foundry UI seems to have caused this

Ok, lmk if you need any assistance 👍

ametel01 committed:
fixed failing tests

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
chore(ci): retrigger build

franciszekjob commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,361 @@
+                }
+                // Skip files containing #[test] attributes
+                if let Ok(contents) = fs::read_to_string(f.path()) {
+                    if contents.contains("#[test]") {
+                        return false;
+                    }
```
This will skip a potential file containing both contract and test - I don't think we should do that. 

cc @kkawula @cptartur

franciszekjob commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,361 @@
+                }
+                // Skip test files
+                let path_str = f.path().to_string_lossy();
+                if path_str.contains("tests") {
```
What if contract file name includes `tests` in the name? imo we should check whether the file is directory and the name is equal to `tests`. Wdyt @ametel01 ?

ametel01 replied:
and maybe also check the file is no in `./src`?

ametel01 committed:
Passes UI to verification interfaces

Refactors the verification process to pass the UI object to the verification interfaces.
This allows the verification implementations to output information and warnings directly to the user.

Updates the voyager implementation to skip test directories but keep files in src/ with tests in their name.

ametel01 commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,361 @@
+        }
+
+        if selected.manifest_metadata.license.is_none() {
+            UI::default().println(&WarningMessage::new("License not specified in Scarb.toml"));
```
patch applied

franciszekjob commented on crates/sncast/tests/e2e/verify/walnut.rs:
```diff
@@ -143,18 +144,19 @@ async fn test_failed_verification_contract_address() {
         .stdin("Y");
 
     let output = snapbox.assert().success();
+    UI::default().println(&format!("{:?}", output.get_output()));
```
`ui.println(...)` should be used

ametel01 replied:
In the test?

franciszekjob replied:
Oops, sorry, ignore it.

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

franciszekjob committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Update docs/src/starknet/verify.md

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/voyager.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/voyager.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/voyager.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/mod.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/voyager.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 commented on crates/sncast/src/starknet_commands/verify/mod.rs:
```diff
@@ -49,30 +58,62 @@ pub struct Verify {
+    } = args;
+
+    let free = network.url(&FreeProvider::semi_random());
+    let rpc_url = url.map_or_else(
+        || {
+            let url = if config.url.is_empty() {
+                &free
+            } else {
+                &config.url
+            };
+            Url::parse(url)
+        },
+        Ok,
+    )?;
+    let provider = JsonRpcClient::new(HttpTransport::new(rpc_url.clone()));
```
changed

ametel01 commented on docs/src/starknet/verify.md:
```diff
@@ -48,6 +52,35 @@ command: verify
+
+    You are about to submit the entire workspace code to the third-party verifier at voyager.
+
+    Important: Make sure your project does not include sensitive information like private keys. The snfoundry.toml file will be uploaded. Keep the keystore outside the project to prevent it from being uploaded.
```
that's a mistake from the docs

The current code does NOT collect snfoundry.toml because:

```rust
.filter(|f| {
    if let Some(ext) = f.path().extension() {
        if ext != OsStr::new(CAIRO_EXT) {  // Only .cairo files
            return false;
        }
        // ... additional filtering
    }
    false
})
```

The filter only includes files with .cairo extension, and snfoundry.toml is not explicitly added like Scarb.toml is.

changed docs and logs to

`Important: Make sure your project's Scarb.toml does not include sensitive information like private keys.`

ametel01 commented on CHANGELOG.md:
```diff
@@ -53,9 +53,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Changed
 
+- `sncast verify` no longer defaults to using walnut.
```
applied

ametel01 committed:
addressed pr review points

ametel01 committed:
Update walnut verification tests

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Updates verification error message format

Updates the format of error messages in the sncast verification tests for better readability.

Specifically, it changes "command: verify" to "Command: verify" and adjusts the formatting of success messages in Walnut tests.

ametel01 committed:
removed unused warning

ametel01 committed:
added explicit binding for wiremock

ametel01 committed:
removed starkent spec expected call

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ddoktorski commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,368 @@
+        let cairo_version = self.metadata.app_version_info.cairo.version.clone();
+        let scarb_version = self.metadata.app_version_info.version.clone();
+
+        let mut workspace_packages: Vec<PackageMetadata> = self
+            .metadata
+            .packages
+            .iter()
+            .filter(|&package_meta| self.metadata.workspace.members.contains(&package_meta.id))
+            .cloned()
+            .collect();
```
I am not sure if I understand the purpose of this operation. Shouldn't it be enough to do `self.metadata.workspace.members`?

ametel01 replied:
`self.metadata.workspace.members` is just a `Vec<PackageId>`, but we need `Vec<PackageMetadata>` for the operations that follow - like checking length, finding by name, etc.

The filtering gets the complete metadata for only workspace packages, excluding dependencies. We can't use just the member IDs since the subsequent code needs the full package metadata.

ddoktorski commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,368 @@
+        match response.status() {
+            StatusCode::OK => {
+                let message = format!(
+                    "{} submitted for verification, you can query the status at: {}/class-verify/job/{}",
```
nit: Please consider extracting endpoints to consts, for example:
```rust
const STATUS_ENDPOINT: &str = "/class-verify/job"
```

ddoktorski commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,369 @@
+                .try_collect()?,
+        };
+
+        let url = format!("{}/class-verify/{:#068x}", self.gen_explorer_url(), hash);
```
> The 68-character format works fine and gives consistent formatting, even though it's not the minimal representation.

Sure, though may I ask why it's not set to 64 if that value is sufficient?

ddoktorski commented on crates/sncast/tests/e2e/verify/voyager.rs:
```diff
@@ -0,0 +1,496 @@
+        .await;
+
+    let job_id = "2b206064-ffee-4955-8a86-1ff3b854416a";
+    let class_hash: Felt =
```
nit:
```suggestion
    let class_hash =
```
Also in other places

ametel01 committed:
Merge branch 'master' into voyager-verify-review-ready

ametel01 committed:
Update Cargo.toml

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,368 @@
+    pub files: HashMap<String, String>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+pub struct ApiError {
+    error: String,
+}
```
changes applied

ametel01 committed:
PR review: Improves `sncast verify` Voyager integration

Enhances the `sncast verify` command's integration with the Voyager block explorer.
The command no longer defaults to Walnut and uses dedicated API endpoints for verification and status checks.
Also improves error reporting.

ametel01 committed:
Merge branch 'voyager-verify-review-ready' of github.com:ametel01/starknet-foundry into voyager-verify-review-ready

franciszekjob commented on CHANGELOG.md:
```diff
@@ -18,10 +18,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 #### Added
 
 - `sncast utils serialize` command to serialize Cairo expressions into calldata
+- `sncast verify` now supports verifying against [voyager](https://voyager.online/) block explorer.
+
```
```suggestion
- `sncast verify` now supports verifying against [voyager](https://voyager.online/) block explorer.
```

franciszekjob commented on crates/sncast/tests/e2e/verify/walnut.rs:
```diff
@@ -145,18 +145,19 @@ async fn test_failed_verification_contract_address() {
             r"
-        Command: verify
-        Error: {}
+        Success: Verification completed
+
+        {}
```
Why does test with failure scenario (`test_failed_verification_contract_address`) results in success?

ametel01 replied:
The walnut verifier now properly handles HTTP error responses by:
  1. Returning errors for non-200 status codes (consistent with voyager)
  2. Displaying error messages in stderr (consistent with voyager)

franciszekjob commented on crates/sncast/tests/e2e/verify/walnut.rs:
```diff
@@ -196,18 +197,19 @@ async fn test_failed_verification_class_hash() {
 
-    assert_stderr_contains(
+    assert_stdout_contains(
         output,
         formatdoc!(
             r"
-        Command: verify
-        Error: {}
+        Success: Verification completed
+
+        {}
```
Same as above.

ametel01 committed:
Handles errors returned by the verifier

Changes the Walnut verification interface to return an error when the verifier responds with an error message.

This ensures that `sncast` correctly handles failed verifications and displays the error message to the user instead of printing a misleading "Success" message.

The tests are updated to reflect the correct error handling.

ddoktorski commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,381 @@
+    Ok(())
+}
+
+fn package_sources(package_metadata: &PackageMetadata) -> Result<Vec<Utf8PathBuf>> {
```
nit:
```suggestion
fn package_source_files(package_metadata: &PackageMetadata) -> Result<Vec<Utf8PathBuf>> {
```

ddoktorski commented on crates/sncast/src/starknet_commands/verify/voyager.rs:
```diff
@@ -0,0 +1,381 @@
+        let mut files: HashMap<String, Utf8PathBuf> = sources
+            .iter()
+            .map(|p| -> Result<(String, Utf8PathBuf)> {
+                let name = p
+                    .strip_prefix(&prefix)
+                    .map_err(|_| anyhow!("Couldn't strip {prefix} from {p}"))?;
+                Ok((name.to_string(), p.clone()))
```
Why store both names and paths rather than just paths?

ametel01 replied:
The relative name is used as the key in the final Body.files field that gets sent to the API.
 The full path is needed to read the file contents with `fs::read_to_string(path.as_path())?`

ddoktorski commented on crates/sncast/tests/e2e/verify/voyager.rs:
```diff
@@ -0,0 +1,488 @@
+    let class_hash = Felt::from_hex(MAP_CONTRACT_CLASS_HASH_SEPOLIA).expect("Invalid class hash");
+
+    Mock::given(method("POST"))
+        .and(path(format!("class-verify/{class_hash:#068x}")))
```
Let's change it in tests as well
```suggestion
        .and(path(format!("class-verify/{class_hash:#064x}")))
```

ametel01 committed:
Update docs/src/starknet/verify.md

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/tests/e2e/verify/voyager.rs

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/src/starknet_commands/verify/voyager.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Update crates/sncast/tests/e2e/verify/voyager.rs

Co-authored-by: ddoktorski <45050160+ddoktorski@users.noreply.github.com>

ametel01 committed:
Allows specifying RPC URL without a network

Updates the `verify` command to allow users to provide an RPC URL directly without specifying a network.
If a URL is provided without a network, defaults to Sepolia.
Adds error handling to ensure that either a network or a URL is provided.
Fixes a typo in the voyager verification command.
Adds a test case to ensure an error is returned when neither network nor URL is provided.

ametel01 commented on crates/sncast/tests/e2e/verify/voyager.rs:
```diff
@@ -0,0 +1,488 @@
+        "--verifier",
+        "voyager",
+        "--network",
+        "sepolia",
```
Applied changes to have --network optional and throw error when --network nor --url are provided, also added a test to verify error triggers correctly.

ddoktorski reviewed (approved):
@ametel01 
Thank you for your contribution!
